examples/shopping-cart/store/modules/cart.js

```diff
@@ -34,41 +33,53 @@ const getters = {
 const actions = {
   checkout ({ commit, state }, products) {
     const savedCartItems = [...state.added]
-    commit(types.SET_CHECKOUT_STATUS, null)
+    commit('setCheckoutStatus', null)
     // empty cart
-    commit(types.SET_CART_ITEMS, { items: [] })
+    commit('setCartItems', { items: [] })
     shop.buyProducts(
       products,
-      () => commit(types.SET_CHECKOUT_STATUS, 'successful'),
+      () => commit('setCheckoutStatus', 'successful'),
       () => {
-        commit(types.SET_CHECKOUT_STATUS, 'failed')
+        commit('setCheckoutStatus', 'failed')
         // rollback to the cart saved before sending the request
-        commit(types.SET_CART_ITEMS, { items: savedCartItems })
+        commit('setCartItems', { items: savedCartItems })
       }
     )
+  },
+
+  addProductToCart ({ state, commit }, product) {
+    commit('setCheckoutStatus', null)
+    if (product.inventory > 0) {
+      const cartItem = state.added.find(item => item.id === product.id)
+      if (!cartItem) {
+        commit('pushProductToCart', { id: product.id })
+      } else {
+        commit('incrementItemQuantity', cartItem)
+      }
+      // remove 1 item from stock
+      commit('decrementProductInventory', { id: product.id })
+    }
   }
 }
 
 // mutations
 const mutations = {
-  [types.ADD_TO_CART] (state, { id }) {
-    state.checkoutStatus = null
-    const record = state.added.find(product => product.id === id)
-    if (!record) {
-      state.added.push({
-        id,
-        quantity: 1
-      })
-    } else {
-      record.quantity++
-    }
+  pushProductToCart (state, { id }) {
+    state.added.push({
+      id,
+      quantity: 1
+    })
+  },
+
+  incrementItemQuantity (state, cartItem) {
+    cartItem.quantity++
```

Comment:
@ktsn what do you think of updating the `cartItem` this way, instead of passing the `id` and looking for it in the state?

Reply:
@hootlex I prefer looking for the target item in the mutation as the example. Because the users can pass a non-state object unexpectedly in this way.

Reply:
That's a good point!

Reply:
@ktsn updated accordingly and used a variable to find the item before updating to demonstrate that normal vars can be used to alter the state.